### PR TITLE
Table page re-design

### LINKED ISSUE #1659: Implement new design for Table Page
- Blocked by #1642

### PR DESCRIPTION
**Pull request by rajatvijay**
Fixes #1659 

Separate issues
- [ ] #2010
- [ ] #2011
- [ ] #2012
- [ ] #2013
- [ ] #2014

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by kgodey** [comment]
I did a quick local review and found a bunch of visual issues. I did not do a thorough review, this is just what I noticed at first glance.

### Badges look weirdly tall
Screen Shot 2022-12-05 at 1 48 44 AM

### The "summarize table" prompt is not dismissable
Screen Shot 2022-12-05 at 1 49 01 AM

### The record summary color does not match Figma
Screen Shot 2022-12-05 at 1 49 04 AM

vs.

Screen Shot 2022-12-05 at 1 50 50 AM

### Grouping badges seem mis-aligned

Screen Shot 2022-12-05 at 1 49 30 AM

### Table icon does not have background color and font formatting seems off
Screen Shot 2022-12-05 at 1 54 59 AM

vs.

Screen Shot 2022-12-05 at 1 50 42 AM

**Comment by kgodey** [comment]
The grouping header style overall seems much less polished than the Figma design. The borders, text sizes, and alignment do not seem correct.

File: mathesar_ui/src/components/TableNameAndDescription.svelte [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+      -webkit-box-orient: vertical;
+      overflow: hidden;
+    }
+
+    .description {
+      display: -webkit-box;
+      -webkit-line-clamp: 1;
+      -webkit-box-orient: vertical;
+      overflow: hidden;
+      font-size: smaller;
```
Comment (seancolsen) [inline mathesar_ui/src/components/TableNameAndDescription.svelte:35]: What's going on with all this vendor-prefixed CSS here? Why is this needed? I'm confused because I wouldn't expect we'd need to use any vendor prefixing here -- and if we did have vendor-prefixed CSS, I'd expect to also see the standard CSS alongside it.

**Review by seancolsen — CHANGES_REQUESTED** [review]
This is looking good! In addition to what Kriti said, I have a few other comments:

- The "Inspector" button should be stylized to indicate the inspector visibility state, as show in Figma.

- Column headers need to display icons at the right, indicating when filtering/sorting/grouping is applied to the table via that column.

- New row indicator does not extend full width
    
    image

- The filtering/sorting/grouping components have a pattern that I don't really like. I can see that this pattern didn't originate within the PR, so apologies if some of this critique is misdirected, but I think it might still be useful to discuss. I'll describe my thoughts in terms of "sorting", though they apply mostly the same for filtering and grouping...

    Let's look at `Sort.svelte` and `SortEntries.svelte`.

    - First of all `SortEntries.svelte` would be better named `SortEntry` because it's displaying _one_ sort entry. (I'll continue using the name `SortEntries` here though for clarity.)

    - `SortEntries` informs its parent of changes via events. I think it would be cleaner to pass the `sorting` store down into `SortEntries`, letting `SortEntries` directly update the store. With the current approach, the updates happen through events which are not type-safe. In `Sort.svelte`, you'll see that `e.detail` has type `any`. This is not good. We can make it type-safe by supplying a type parameter to `createEventDispatcher`, as is done in `FilterEntries`, but I prefer passing the store down into the lower-level components because it's simpler. In cases where there is looser coupling between the parent/child component, then I think it can make sense to delegate more business logic to the parent (thereby creating the sort of "dumb" child components that I think you're after here), but in this case we have components that are inherently tightly coupled anyway, so I'd rather embrace that and put the business logic closest to its relevant UI.

    This is not a huge deal though, and I don't think you necessarily need to make any changes in this PR. At the very least, for now it would be nice to modify `SortEntries.svelte` adding a type parameter to `createEventDispatcher` that ensures `e.detail` is typed.

I also opened some tickets for more changes to make later:

- The UI that displays the loading and saved status (e.g. "Saving changes", "Couldn't save changes", etc) does not match Figma. But I think that's good, actually, because I've been working on building that UI for the Record Page as part of #1976. I have a `ModificationStatus` component in that PR that I'm hoping we can use on the Table Page too. To track that task, I've opened #2021 and assigned it to you.

- #2020

- #2022

**Comment by rajatvijay** [comment]
> This is looking good! In addition to what Kriti said, I have a few other comments:
> 
>     * The "Inspector" button should be stylized to indicate the inspector visibility state, as show in Figma.

66227175a

>     * Column headers need to display icons at the right, indicating when filtering/sorting/grouping is applied to the table via that column.

a0c2ca3ef

>     * New row indicator does not extend full width, and (if you look _very_ closely), you'll see that a border is overlapping the text on the right side
>       image

23786c4e7


>     * The filtering/sorting/grouping components have a pattern that I don't really like. I can see that this pattern didn't originate within the PR, so apologies if some of this critique is misdirected, but I think it might still be useful to discuss. I'll describe my thoughts in terms of "sorting", though they apply mostly the same for filtering and grouping...
>       Let's look at `Sort.svelte` and `SortEntries.svelte`.
>       
>       * First of all `SortEntries.svelte` would be better named `SortEntry` because it's displaying _one_ sort entry. (I'll continue using the name `SortEntries` here though for clarity.)
>       * `SortEntries` informs its parent of changes via events. I think it would be cleaner to pass the `sorting` store down into `SortEntries`, letting `SortEntries` directly update the store. With the current approach, the updates happen through events which are not type-safe. In `Sort.svelte`, you'll see that `e.detail` has type `any`. This is not good. We can make it type-safe by supplying a type parameter to `createEventDispatcher`, as is done in `FilterEntries`, but I prefer passing the store down into the lower-level components because it's simpler. In cases where there is looser coupling between the parent/child component, then I think it can make sense to delegate more business logic to the parent (thereby creating the sort of "dumb" child components that I think you're after here), but in this case we have components that are inherently tightly coupled anyway, so I'd rather embrace that and put the business logic closest to its relevant UI.
>       
>       This is not a huge deal though, and I don't think you necessarily need to make any changes in this PR. At the very least, for now it would be nice to modify `SortEntries.svelte` adding a type parameter to `createEventDispatcher` that ensures `e.detail` is typed.
> 

bf2e67d9f

> I also opened some tickets for more changes to make later:
> 
>     * The UI that displays the loading and saved status (e.g. "Saving changes", "Couldn't save changes", etc) does not match Figma. But I think that's good, actually, because I've been working on building that UI for the Record Page as part of #1976. I have a `ModificationStatus` component in that PR that I'm hoping we can use on the Table Page too. To track that task, I've opened #2021 and assigned it to you.
> 
>     * #2020
> 
>     * #2022

**Comment by rajatvijay** [comment]
> I did a quick local review and found a bunch of visual issues. I did not do a thorough review, this is just what I noticed at first glance.
> ### Badges look weirdly tall
> Screen Shot 2022-12-05 at 1 48 44 AM

448a4f430

> ### The "summarize table" prompt is not dismissable
> Screen Shot 2022-12-05 at 1 49 01 AM

https://github.com/centerofci/mathesar/issues/2024

> ### The record summary color does not match Figma
> Screen Shot 2022-12-05 at 1 49 04 AM
> 
> vs.
> 
> Screen Shot 2022-12-05 at 1 50 50 AM

ef4838f53

> ### Grouping badges seem misaligned
> 
> Screen Shot 2022-12-05 at 1 49 30 AM

a87bb0b58

> ### Table icon does not have background color and font formatting seems off
> 
> Screen Shot 2022-12-05 at 1 54 59 AM
> 
> vs.
> 
> Screen Shot 2022-12-05 at 1 50 42 AM

We do not support colors for tables as of now. This is another one of those discussions that I had with @ghislaineguerin to not do it for now. Created an issue for this here: https://github.com/centerofci/mathesar/issues/2025

File: mathesar_ui/src/components/TableNameAndDescription.svelte (hunk 1 above)
Comment (seancolsen) [reply mathesar_ui/src/components/TableNameAndDescription.svelte:34]: Rajat and I discussed this on a call. We're leaving the code as-is.

**Comment by rajatvijay** [comment]
> Everything looks good except one thing:
> 
>     * The group header row still doesn't quite look as good as in Figma. In Figma, it starts further over to the left and has a bit more left padding.
>       
>       
>       
>       Figma
>       PR
>       
>       
>       
>       
>       image
>       image

Apologies for this, 10931302b

**Review by seancolsen — APPROVED** [review]
Looks good!

@kgodey, Rajat seems to have addressed all your critique too, so I'm merging this.
